hare/algorithm.go

```diff
@@ -88,140 +101,505 @@ func (proc *ConsensusProcess) createInbox(size uint32) chan *pb.HareMessage {
+		}
 	}
+
+	// done with building proposal, reset statuses tracking
+	proc.statusesTracker = nil
 }
 
-// TODO: maybe change it to SetStatusMessage and just set proc.roundMsg?
-func (proc *ConsensusProcess) buildStatusMessage() ([]byte, error) {
-	x := NewInnerBuilder().SetType(Status).SetLayer(proc.layerId).SetIteration(proc.k).SetKi(proc.ki).SetBlocks(proc.s).Build()
+func (proc *ConsensusProcess) beginRound3() {
+	proposedSet := proc.proposalTracker.ProposedSet()
+	if proposedSet != nil { // has proposal to send
+		builder := proc.initDefaultBuilder(proposedSet).SetType(Commit).Sign(proc.signing)
+		roundMsg := builder.Build()
+		proc.sendMessage(roundMsg)
+	}
 
-	buff, err := proto.Marshal(x)
-	if err != nil {
-		return nil, errors.New("error marshaling inner message")
+	// proposedSet may be nil, in such case the tracker will ignore messages
+	proc.commitTracker = NewCommitTracker(proc.cfg.F+1, proc.cfg.N, proposedSet) // track commits for proposed set
+}
+
+func (proc *ConsensusProcess) beginRound4() {
+	proc.commitTracker = nil
+	proc.proposalTracker = nil
+}
+
+func (proc *ConsensusProcess) onRoundBegin() {
+	proc.updateRole()
+
+	// reset trackers
+	switch proc.currentRound() {
+	case Round1:
+		proc.beginRound1()
+	case Round2:
+		proc.beginRound2()
+	case Round3:
+		proc.beginRound3()
+	case Round4:
+		proc.beginRound4()
+	default:
+		log.Error("Current round out of bounds. Expected: 0-4, Found: ", proc.currentRound())
+		panic("Current round out of bounds")
 	}
+}
 
-	sig := proc.signing.Sign(buff)
+func (proc *ConsensusProcess) roleProof() Signature {
+	kInBytes := make([]byte, 4)
+	binary.LittleEndian.PutUint32(kInBytes, uint32(proc.k))
 
-	outer := NewOuterBuilder()
-	m := outer.SetInnerMessage(x).SetPubKey(proc.pubKey).SetInnerSignature(sig).Build()
+	return proc.signing.Sign(kInBytes)
+}
 
-	buff, err = proto.Marshal(m)
-	if err != nil {
-		return nil, errors.New("error marshaling message")
+func (proc *ConsensusProcess) initDefaultBuilder(s *Set) *MessageBuilder {
+	builder := NewMessageBuilder().SetPubKey(proc.pubKey).SetInstanceId(proc.instanceId)
+	builder = builder.SetRoundCounter(proc.k).SetKi(proc.ki).SetValues(s)
+	builder.SetRoleProof(proc.roleProof())
+
+	return builder
+}
+
+func (proc *ConsensusProcess) buildNotifyMessage() *pb.HareMessage {
+	builder := proc.initDefaultBuilder(proc.s).SetType(Notify).SetCertificate(proc.certificate).Sign(proc.signing)
+
+	return builder.Build()
+}
+
+func (proc *ConsensusProcess) processPreRoundMsg(msg *pb.HareMessage) {
+	proc.preRoundTracker.OnPreRound(msg)
+}
+
+func (proc *ConsensusProcess) processStatusMsg(msg *pb.HareMessage) {
+	s := NewSet(msg.Message.Values)
+	if proc.preRoundTracker.CanProveSet(s) {
+		proc.statusesTracker.RecordStatus(msg)
+	}
+}
+
+// validate proposal for type A of the proposal (where ki=-1)
+func (proc *ConsensusProcess) validateProposalTypeA(m *pb.HareMessage) bool {
+	s := NewSet(m.Message.Values)
+	unionSet := NewEmptySet(proc.cfg.SetSize)
```

Comment:
Not exactly the same. One iterates a map and the other an array. I noticed that before, but since it is a little piece of code I think it is acceptable. ("A little copying is better than a little dependency", Rob)